### PR TITLE
NEW email notification for MFA settings update

### LINKED ISSUE #67: Changes to user's MFA should result in email
As a user with a registered MFA method, I would like to receive email notifications for all updates related to my MFA, so that if I receive an email referencing a change I am not aware of I can expect my account to be compromised and follow the recommended steps. 

**ACs**
- Any actions changing or resetting authentication mechanisms will trigger automated emails to the email address of the user
- Email should make specific reference to the MFA method
- Email should provide steps for the user to take if the change was not expected
- Email template is theme-able 

**Notes**
- Examples of changing or reseting:
a. MFA method is registered
b. MFA method is removed
c. Recovery codes are reset
d. Recovery code is used
e. One recovery code left?

### PR DESCRIPTION
**Pull request by NightJar**
As a precaution users should be notified whenever their account details pertaining to security are updated, in order to let them respond to what could be a security breach. The easiest manner to do this is to send an email to their registered address.

Closes #67

### CODE REVIEW

File: src/Service/Notification.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+     *
+     * @return void
+     */
+    public function send()
+    {
+        $this->extend('onBeforeSend', $this->email);
+        try {
+            $this->email->send();
+        } catch (Exception $e) {
+            Injector::inst()->get(LoggerInterface::class . '.MFA')->info($e->getMessage());
```
Comment (robbieaverill) [inline src/Service/Notification.php]: I think it'd be good to put this into DI configuration if you can. Also, do you want an `onAfterSend` hook as well? Currently you could put some logic into `onBeforeSend` that would fire even if the email send fails, but you may want to hook into it and check the result of the send operation

File: templates/Email/MFA/Notification_backupcodeused.ss [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<p><%t SilverStripe\\MFA\\Email\\Notification.HELLO 'Hi' %> $Member.Name,</p>
+
+<p>
+	<%t SilverStripe\\MFA\\Email\\Notification.USEDBACKUPCODE 'A recovery code was used to gain access to your account at' %> $AbsoluteBaseURL.<br />
```
Comment (robbieaverill) [inline templates/Email/MFA/Notification_backupcodeused.ss]: * It's probably worth including the link as part of the translation, in case RTL languages need to put it first
* I think the <br> could just be two separate paragraph tags

File: templates/Email/MFA/Notification_backupcodeused.ss [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+	<%t SilverStripe\\MFA\\Email\\Notification.CODESLEFT 'You now have {count} codes remaining.' count=$CodesRemaining %>
+</p>
+
+<p>
+    <%t SilverStripe\\MFA\\Email\\Notification.NOTYOU 'If you did not take this action, please contact your system administrator immediately' %>
```
Comment (robbieaverill) [inline templates/Email/MFA/Notification_backupcodeused.ss]: Can you ensure each sentence has a full stop at the end?

File: tests/php/BackupCode/VerifyHandlerTest.php [hunk 4]
```diff
@@ -17,6 +18,18 @@ class VerifyHandlerTest extends SapphireTest
 {
     protected static $fixture_file = 'VerifyHandlerTest.yml';
 
+    public function setUp()
```
Comment (robbieaverill) [inline tests/php/BackupCode/VerifyHandlerTest.php]:
```suggestion
    protected function setUp()
```

File: tests/php/BackupCode/VerifyHandlerTest.php [hunk 5]
```diff
@@ -17,6 +18,18 @@ class VerifyHandlerTest extends SapphireTest
...
+    {
+        parent::setUp();
+
+        //Ensure that the handler for notifications is Email - which is stubbed with Swift_NullTransport in test state
+        Injector::inst()->load([
+            Email::class . '.MFA.Notification' => [
```
Comment (robbieaverill) [inline tests/php/BackupCode/VerifyHandlerTest.php]: Does this actually work in practice? I notice that you create the service definition in this test, but it's not defined in practice. Does the class that loads `Email.MFA.Notification` actually work without it being defined?

File: tests/php/Service/RegisteredMethodManagerTest.php [hunk 6]
```diff
@@ -23,6 +25,18 @@ class RegisteredMethodManagerTest extends SapphireTest
         ],
     ];
 
+    public function setUp()
```
Comment (robbieaverill) [inline tests/php/Service/RegisteredMethodManagerTest.php]:
```suggestion
    protected function setUp()
```

File: tests/php/Service/RegisteredMethodManagerTest.php [hunk 7]
```diff
@@ -23,6 +25,18 @@ class RegisteredMethodManagerTest extends SapphireTest
...
+    {
+        parent::setUp();
+
+        //Ensure that the handler for notifications is Email - which is stubbed with Swift_NullTransport in test state
+        Injector::inst()->load([
+            Email::class . '.MFA.Notification' => [
+                'class' => Email::class
+            ],
+        ]);
```
Comment (robbieaverill) [inline tests/php/Service/RegisteredMethodManagerTest.php]: SapphireTest configures a TestMailer implementation by default. Do you need this?

File: tests/php/Service/RegisteredMethodManagerTest.php [hunk 8]
```diff
@@ -66,6 +80,33 @@ public function testRegisterForMemberCreatesNewMethod()
...
+    {
+        /** @var Member&MemberExtension $member */
+        $member = Member::create(['FirstName' => 'Mike', 'Email' => 'test@example.com']);
+        $member->write();
+        $method = new BackupCodeMethod();
+
+        $manager = RegisteredMethodManager::singleton();
+        RegisteredMethodManager::singleton()->registerForMember($member, $method, ['foo', 'bar']);
+
+        $this->assertNull(static::findEmail($member->Email));
```
Comment (robbieaverill) [inline tests/php/Service/RegisteredMethodManagerTest.php]:
```suggestion
        $this->assertNull($this->findEmail($member->Email));
```

File: tests/php/Service/RegisteredMethodManagerTest.php [hunk 9]
```diff
@@ -93,6 +134,17 @@ public function testDeleteFromMember()
...
 
+    public function testDeleteFromMemberSendsNotification()
+    {
+        /** @var Member&MemberExtension $member */
+        $member = $this->objFromFixture(Member::class, 'bob_jones');
+
+        $manager = RegisteredMethodManager::singleton();
+        $manager->deleteFromMember($member, new BasicMathMethod());
+
+        $this->assertEmailSent($member->Email);
```
Comment (robbieaverill) [inline tests/php/Service/RegisteredMethodManagerTest.php]: It might be worth asserting some of the subject line to ensure the correct email was sent - you can pass regex into the arguments e.g. `$this->assertEmailSend($member->Email, null, '/method was deleted/')`

File: tests/php/BackupCode/VerifyHandlerTest.php (hunk 5 above)
Comment (NightJar) [reply tests/php/BackupCode/VerifyHandlerTest.php]: Yes, it works as intended :)

Comment (robbieaverill) [reply tests/php/BackupCode/VerifyHandlerTest.php]: Intended meaning it falls back up to `...Email.MFA` when `...Email.MFA.NonExistentService` doesn't exist?

File: _config/notifications.yml [hunk 10]
```diff
@@ -0,0 +1,3 @@
+SilverStripe\Core\Injector\Injector:
+  SilverStripe\Control\Email\Email.MFA:
```
[the remark this answers is not in the document]
Comment (NightJar) [reply _config/notifications.yml]: ok

File: src/Service/Notification.php [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+     */
+    public function __construct(Member $member, string $template, array $data = [])
+    {
+        $this->email = Injector::inst()->create(Email::class . '.MFA.Notification')
+            ->setTo($member->Email)
+            ->setHTMLTemplate($template)
+            ->setData(array_merge(['Member' => $member], $data));
+        foreach (['to', 'from', 'subject'] as $header) {
+            if (isset($data[$header])) {
+                $method = 'set' . ucwords($header);
```
[the remark this answers is not in the document]
Comment (NightJar) [reply src/Service/Notification.php]: I don't think so, it doesn't look like swift mailer handles null inputs (it gets passed straight through).

[the remark this answers is not in the document]
Comment (robbieaverill) [reply src/Service/Notification.php]: To be fair, none of these arguments sound optional to me...

File: templates/Email/MFA/Notification_backupcodeused.ss (hunk 2 above)
Comment (NightJar) [reply templates/Email/MFA/Notification_backupcodeused.ss]: hahaha markdown converts HTML 😂

File: tests/php/Service/RegisteredMethodManagerTest.php (hunk 7 above)
Comment (NightJar) [reply tests/php/Service/RegisteredMethodManagerTest.php]: Yes, it is possible to configure `Email::class` to be a service other than `Email::class` - which `SapphireTest` will rectify (maybe, I didn't look)...
However through the magic of Injector the dot context is more specific and allows people to override this only in the `MFA.Notification` context, which will in turn override whatever `SapphireTest` configures.

Comment (robbieaverill) [reply tests/php/Service/RegisteredMethodManagerTest.php]: Ah good point. And what's the reasoning for defining a sub-service of this? Are you expecting people to want to use a different mailer service for MFA emails?

File: tests/php/Service/RegisteredMethodManagerTest.php (hunk 8 above)
Comment (NightJar) [reply tests/php/Service/RegisteredMethodManagerTest.php]: this is a static method

Comment (robbieaverill) [reply tests/php/Service/RegisteredMethodManagerTest.php]: Yeah, so are all of PHPUnit's assertions, but that doesn't mean you have to call them statically

Comment (robbieaverill) [reply tests/php/Service/RegisteredMethodManagerTest.php]: BTW the reason I mentioned this is because `assertEmailSent()` is static and you're calling it non-statically in the same class

File: tests/php/Service/RegisteredMethodManagerTest.php (hunk 9 above)
Comment (NightJar) [reply tests/php/Service/RegisteredMethodManagerTest.php]: Are tests always run in English?

Comment (robbieaverill) [reply tests/php/Service/RegisteredMethodManagerTest.php]: We can assume they are, yes. SapphireTest sets the locale to the configured default locale which is en_US by default. If you want to be super safe you can add the same translation calls in your assertions as in the source class, but Guy's actually proposing to remove non-English translations from unit tests by default in an upcoming release. We don't usually bother with this in assertions, just check English

File: tests/php/Service/RegisteredMethodManagerTest.php (hunk 8 above)
Comment (NightJar) [reply tests/php/Service/RegisteredMethodManagerTest.php]: Oh 😅

File: tests/php/Service/RegisteredMethodManagerTest.php (hunk 7 above)
Comment (NightJar) [reply tests/php/Service/RegisteredMethodManagerTest.php]: Yeah, potentially. I don't think extra context is a bad thing - one could configure e.g. a custom emailer for most (mfa) things, and an SMS service for notifications only.
It's quite unlikely I'll admit, but nor does the extra context do any harm as the fall back happens regardless (as noted by the service defined in `notifications.yml` added in this same commit).

The only reason that is defined at all (to what is otherwise the default for that class ID) is because we need to support all versions of SS4, so when https://github.com/silverstripe/silverstripe-framework/pull/8634 is finally updated with unit tests this will still be necessary for 4.0 -> 4.5+ (or whenever it actually gets merged).

File: tests/php/BackupCode/VerifyHandlerTest.php (hunk 5 above)
Comment (NightJar) [reply tests/php/BackupCode/VerifyHandlerTest.php]: Yup, exactly :)

File: src/Service/Notification.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (NightJar) [reply src/Service/Notification.php]: They all are except `to` - which is set to the member by default, which is a non-optional argument.

File: tests/php/Service/RegisteredMethodManagerTest.php (hunk 9 above)
Comment (NightJar) [reply tests/php/Service/RegisteredMethodManagerTest.php]: OK neat :)

File: src/Service/Notification.php (hunk 1 above)
Comment (NightJar) [reply src/Service/Notification.php]: I added the after hook, but am not sure what you mean by the DI configuration part.
We already have a DI configuration for MFA logging, however I did have the case wrong on the context (now fixed).

**Comment by NightJar** [comment]
Cool, I adapted to along the lines of your comment too @ScopeyNZ - however I didn't see the point in dependencies as the execution already fetches a named service. Leaving a member variable would still have state, whether it be given or installed via Injector.

File: tests/php/Service/RegisteredMethodManagerTest.php (hunk 9 above)
Comment (NightJar) [reply tests/php/Service/RegisteredMethodManagerTest.php]: Although this isn't probably the best place to continue this chat...
The default translation for a site isn't necessarily English - the second parameter to `_t()` as I understand it is the fallback - it's still possible for a translation file to override this. I think ideally one would have a test state for translation or set to a non-existent lang such as "SS_TEST" in order to trigger the known state (fallback text).

File: tests/php/Service/RegisteredMethodManagerTest.php (hunk 7 above)
Comment (robbieaverill) [reply tests/php/Service/RegisteredMethodManagerTest.php]: I think it's pretty edge casey to be honest, and the tradeoff is that you have to mock the service manually in these tests. If someone else were running a test that inadvertently triggers this logic then the email handler would still be set. You may consider putting this logic into a TestState implementation so it'd be global, but personally I'd be more in favour of just using the default email implementation since we haven't got any other examples of code yet that deviate from it.

File: tests/php/Service/RegisteredMethodManagerTest.php (hunk 9 above)
Comment (robbieaverill) [reply tests/php/Service/RegisteredMethodManagerTest.php]: Correct. You have two options:

* Include identical `_t()` calls in your assertions (we've done this before in unit tests)
* Assume it'll be English (we've done this before in unit tests)

File: tests/php/Service/RegisteredMethodManagerTest.php (hunk 8 above)
Comment (NightJar) [reply tests/php/Service/RegisteredMethodManagerTest.php]: I updated all static calls to be statically called.

**Comment by NightJar** [comment]
Broken build from race condition. Rerun has made them green.

File: src/Service/Notification.php [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+                $email->$method($data[$header]);
+            }
+        }
+
+        $this->extend('onBeforeSend', $email);
+
+        try {
+            $sendResult = $email->send();
+        } catch (Exception $e) {
+            Injector::inst()->get(LoggerInterface::class . '.mfa')->info($e->getMessage());
```
Comment (robbieaverill) [inline src/Service/Notification.php]: Continuing on from https://github.com/silverstripe/silverstripe-mfa/pull/134#discussion_r286759579 which I can't reply to now for some reason, here's an example:

```php
protected $logger;
private static $dependencies = ['Logger' => '%$' . LoggerInterface::class . '.mfa'];

public function setLogger(LoggerInterface $logger) { $this->logger = $logger; return $this; }

// ...

} catch (Exception $e) {
    $this->logger->info($e->getMessage());
}
```

File: src/Service/Notification.php (hunk 1 above)
Comment (robbieaverill) [reply src/Service/Notification.php]: Here's an example:

```php
protected $logger;
private static $dependencies = ['Logger' => '%$' . LoggerInterface::class . '.mfa'];

public function setLogger(LoggerInterface $logger) { $this->logger = $logger; return $this; }

// ...

} catch (Exception $e) {
    $this->logger->info($e->getMessage());
}
```

File: tests/php/Service/RegisteredMethodManagerTest.php (hunk 9 above)
Comment (NightJar) [reply tests/php/Service/RegisteredMethodManagerTest.php]: I've assumed English.

File: tests/php/Service/RegisteredMethodManagerTest.php (hunk 8 above)
Comment (NightJar) [reply tests/php/Service/RegisteredMethodManagerTest.php]: Oh, oh.
Let me change again.

File: templates/Email/MFA/Notification_backupcodeused.ss (hunk 3 above)
Comment (NightJar) [reply templates/Email/MFA/Notification_backupcodeused.ss]: I have done this.

File: tests/php/BackupCode/VerifyHandlerTest.php (hunk 4 above)
Comment (NightJar) [reply tests/php/BackupCode/VerifyHandlerTest.php]: Done.

File: tests/php/Service/RegisteredMethodManagerTest.php (hunk 7 above)
Comment (robbieaverill) [reply tests/php/Service/RegisteredMethodManagerTest.php]: To clarify - I'm suggesting you remove the use of sub-service definitions and use the default mailer

File: templates/Email/MFA/Notification_backupcodeused.ss (hunk 2 above)
Comment (NightJar) [reply templates/Email/MFA/Notification_backupcodeused.ss]: these changes were made.

File: src/Service/RegisteredMethodManager.php [hunk 13]
```diff
@@ -60,6 +60,20 @@ public function registerForMember(Member $member, MethodInterface $method, $data
         // Add it to the member
         $member->RegisteredMFAMethods()->add($registeredMethod);
 
+        if (!MethodRegistry::create()->isBackupMethod($method)) {
+            Notification::create()->send(
```
Comment (robbieaverill) [inline src/Service/RegisteredMethodManager.php]: I think this is where @ScopeyNZ was suggesting that you can use dependency injection for the Notification class as a singleton

File: tests/php/Service/RegisteredMethodManagerTest.php (hunk 7 above)
Comment (NightJar) [reply tests/php/Service/RegisteredMethodManagerTest.php]: Removed context.
`Notification` class is now essentially just:
`Email::create()->setTo($to)->setHTMLTemplate($template)->setData($data)->send()`

If someone wants to use a different service they can Inject `Notification` over `Email.mfa`.

File: src/Service/Notification.php (hunk 1 above)
Comment (NightJar) [reply src/Service/Notification.php]: Did this.

File: tests/php/BackupCode/VerifyHandlerTest.php (hunk 5 above)
Comment (NightJar) [reply tests/php/BackupCode/VerifyHandlerTest.php]: This code is removed.

File: src/Service/RegisteredMethodManager.php (hunk 13 above)
Comment (NightJar) [reply src/Service/RegisteredMethodManager.php]: converted.

File: src/Service/Notification.php [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+     *
+     * Builds the notification delivery class & adds details to it. This is by default an {@see Email} as built by
+     * Injector. Data can contain 'from', 'to', and 'subject' keys, which will be set on the Email before sending.
+     * A caveat to be aware of is that setting 'to' will obviously override the given Member as the recipient.
+     *
+     * @param Member $member Member the notification is being sent to
+     * @param string $template Name of the HTMLTemplate to use for the email body
+     * @param array $data (optional) Extra data for use in populating the template
+     *
+     * @return void
```
Comment (robbieaverill) [inline src/Service/Notification.php]: Do you want to `return $sendResult` (bool) instead?

File: _config/notifications.yml [hunk 15]
```diff
@@ -0,0 +1,6 @@
+---
+Name: mfa-notifications
+---
+SilverStripe\Core\Injector\Injector:
+  SilverStripe\Control\Email\Email.mfa:
+    class: SilverStripe\Control\Email\Email
```
Comment (robbieaverill) [inline _config/notifications.yml]: Do we still need this config?

File: src/Service/RegisteredMethodManager.php (hunk 13 above)
Comment (robbieaverill) [reply src/Service/RegisteredMethodManager.php]: Did you forget to push?

Comment (NightJar) [reply src/Service/RegisteredMethodManager.php]: [No?](https://github.com/silverstripe/silverstripe-mfa/pull/134/commits/20db45104a9b5b1b44aa27619f95f0180700e264#diff-de30d7d32ad168198aa31a778c164447R64)

Comment (robbieaverill) [reply src/Service/RegisteredMethodManager.php]: Simply changing the class from a non-singleton to a singleton isn't implementing dependency injection. Are we treating Notification as a stateless service class? If so, the suggested change would've been like this:

```php
private static $dependencies = ['Notification' => '%$' . Notification::class];
protected $notification;
public function setNotification(Notification $notification) { $this->notification = $notification; return $this; }

/// ...

if (!MethodRegi.....) {
     $this->notification->send(...);
}
```

Note that if this is what we're expecting then I'd also suggest renaming the class to `NotificationService` to indicate that that's what it is rather than an individual model.

**Comment by NightJar** [comment]
addressed and tests passed.

File: src/BackupCode/VerifyHandler.php [hunk 16]
```diff
@@ -58,6 +59,14 @@ public function verify(HTTPRequest $request, StoreInterface $store, RegisteredMe
                 array_splice($candidates, $index, 1);
                 $registeredMethod->Data = json_encode($candidates);
                 $registeredMethod->write();
+                Notification::singleton()->send(
```
Comment (robbieaverill) [inline src/BackupCode/VerifyHandler.php]: DI?

File: src/Service/Notification.php [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+     *
+     * @param Member $member Member the notification is being sent to
+     * @param string $template Name of the HTMLTemplate to use for the email body
+     * @param array $data (optional) Extra data for use in populating the template
+     *
+     * @return bool
+     */
+    public function send(Member $member, string $template, array $data = []): bool
+    {
+        $this->email
```
Comment (robbieaverill) [inline src/Service/Notification.php]: This will make the email object a singleton (by using it in DI) - is that what you expected?

File: src/Service/Notification.php [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+use SilverStripe\Control\Email\Email;
+use SilverStripe\Core\Extensible;
+use SilverStripe\Core\Injector\Injectable;
+use SilverStripe\Security\Member;
+
+/**
+ * Ecapsulates setting up an Email in order to allow for Dependency Injection and to avoid introducing a hard
+ * coupling to the SilverStripe core Email class in code that consumes this class.
+ */
+class Notification
```
Comment (robbieaverill) [inline src/Service/Notification.php]: Did you want to rename the class? The reason I suggested it because `$dependencies = ['NotificationService' => Notification::class]` is not as verbose as `NotificationService::class` despite it being namespaced. If you think it's fine then I'm happy too. You could also use `Notification\Service::class`, or leave it as is. The property name helps 👍

File: src/Service/Notification.php (hunk 17 above)
Comment (NightJar) [reply src/Service/Notification.php]: err, no. thanks for pointing that out.

File: src/Service/Notification.php [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+        try {
+            $sendResult = $email->send();
+        } catch (Exception $e) {
+            $this->logger->info($e->getMessage());
+        }
+
+        $this->extend('onAfterSend', $email, $sendResult);
+
+        // Could no longer be a bool as a result of the extension point above, so is cast to bool.
+        return (bool) $sendResult;
```
Comment (robbieaverill) [inline src/Service/Notification.php]: You could also use a `SilverStripe\MFA\State\Result` object which could provide more context. That was extensions can say why they're changing it to a failure as well. We could also provide `$ex->getMessage()` as `context` when catching exceptions and setting it to be a failure. I think a bool is fine for now but maybe something you might want to follow up on later 👍
